projects/solar/test/test_data_tx.c

```diff
@@ -0,0 +1,120 @@
+    CAN_UNPACK_SOLAR_DATA(msg, (uint32_t *)&s_can_msgs_data_points[s_can_msg_count],
+                          &s_can_msg_values[s_can_msg_count]);
+    s_can_msg_count++;
+    return STATUS_CODE_OK;
+  }
+  return STATUS_CODE_RESOURCE_EXHAUSTED;
+}
+
+// Set every data point
+static void prv_test_data_tx_set_data_values(void) {
+  for (DataPoint data_point = 0; data_point < NUM_DATA_POINTS; data_point++) {
+    TEST_ASSERT_OK(data_store_set(data_point, data_point));
+  }
+}
+
+static void prv_reset_arrays(void) {
+  for (uint16_t index = 0; index < 2 * NUM_DATA_POINTS; index++) {
+    s_can_msg_values[index] = 0;
+    s_can_msgs_data_points[index] = 0;
+  }
+}
+
+static void prv_process_can_events(SoftTimerId timer_id, void *context) {
+  for (uint16_t msg = 0; msg < s_data_tx_settings.msgs_per_tx_iteration; msg++) {
+    MS_TEST_HELPER_CAN_TX(SOLAR_CAN_EVENT_TX);
+  }
+  for (uint16_t msg = 0; msg < s_data_tx_settings.msgs_per_tx_iteration; msg++) {
+    MS_TEST_HELPER_CAN_RX(SOLAR_CAN_EVENT_RX);
+  }
+  s_num_callbacks++;
+  if (s_can_msg_count < NUM_DATA_POINTS)
+    TEST_ASSERT_OK(soft_timer_start_millis(s_data_tx_settings.wait_between_tx_in_millis,
+                                           prv_process_can_events, NULL, NULL));
+}
+
+void setup_test(void) {
+  TEST_ASSERT_OK(initialize_can_and_dependencies(&s_can_storage, SYSTEM_CAN_DEVICE_SOLAR,
+                                                 SOLAR_CAN_EVENT_TX, SOLAR_CAN_EVENT_RX,
+                                                 SOLAR_CAN_EVENT_FAULT));
+  TEST_ASSERT_OK(can_register_rx_handler(SYSTEM_CAN_MESSAGE_SOLAR_DATA,
+                                         prv_test_data_tx_callback_handler, NULL));
+  TEST_ASSERT_OK(data_store_init());
+
+  TEST_ASSERT_OK(data_tx_init(&s_data_tx_settings));
+
+  s_can_msg_count = 0;
+  s_num_callbacks = 0;
+  prv_reset_arrays();
+}
+
+void teardown_test(void) {}
+
+// ensure no can messages are sent when data points are not set
+void test_data_tx_no_values_set(void) {
+  Event e = { .id = DATA_READY_EVENT };
+  TEST_ASSERT(data_tx_process_event(&e));
+
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+}
+
+// ensure data_tx_process_event can handle NULL events or irrelevant events
+void test_data_tx_invalid_event(void) {
+  TEST_ASSERT_FALSE(data_tx_process_event(NULL));
+  Event e = { .id = DATA_READY_EVENT + 1 };
+  TEST_ASSERT_FALSE(data_tx_process_event(&e));
+}
+
+// set all data points, tx all data points, verify the values of the data point and value in the
+// expected order
+void test_data_tx(void) {
+  prv_test_data_tx_set_data_values();
+  Event e = { .id = DATA_READY_EVENT };
+  TEST_ASSERT(data_tx_process_event(&e));
+
+  prv_process_can_events(SOFT_TIMER_INVALID_TIMER, NULL);
+  int32_t delay_time = (NUM_DATA_POINTS / (int32_t)s_data_tx_settings.msgs_per_tx_iteration + 1) *
+                       (int32_t)s_data_tx_settings.wait_between_tx_in_millis;
+  delay_ms((uint32_t)delay_time);
```

Comment:
Why all the casts with `int32_t`? If it's because `NUM_DATA_POINTS` is signed, I'd prefer you cast it instead, and if it's to widen everything to 32 bits I'd prefer you cast to `uint32_t` instead. Not a super big deal though.